### PR TITLE
Jackson: Do not serialize empty error and dependency collections

### PR DESCRIPTION
**Pull request by mnonnenmacher**
Do not serialize empty error or dependency collections. This improves the
readability of the output files and saves disk space.

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/939)

### CODE REVIEW

File: analyzer/src/funTest/assets/projects/synthetic/gradle-all-dependencies-expected-result-with-curations.yml [hunk 1]
```diff
@@ -676,6 +537,5 @@ analyzer:
           - "curated license a"
           - "curated license b"
           comment: "Declared license in pom.xml is wrong."
-    errors: {}
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/gradle-all-dependencies-expected-result-with-curations.yml]: Only semi-related, but I'm wondering: Were we actually ever populating this `errors` map? Even for files like this one where we actually do have errors in the tree, this is empty. Also a quick search through the code indicates we never write to this field. If so, we should probably remove it from the `AnalyzerResult` class in a preceeding commit.

File: model/src/main/kotlin/ProjectAnalyzerResult.kt [hunk 2]
```diff
@@ -39,6 +41,7 @@ data class ProjectAnalyzerResult(
         /**
          * The list of errors that occurred during dependency resolution. Defaults to an empty list.
          */
+        @JsonInclude(JsonInclude.Include.NON_EMPTY)
```
Comment (sschuberth) [inline model/src/main/kotlin/ProjectAnalyzerResult.kt]: I believe we're never ever (de-)serializing this class directly, so we *could* drop this.

**Comment by sschuberth** [comment]
As discussed offline I actually *do* like the direction of this change. However, it would make me feel better if we could come up with a more clear rule for ourselves to better explain which empty fields we do serialize, and which ones not.

For example, why are empty `curations` and `license_findings` still serialized? But I'm not saying they should not.

I guess my proposal would be to only omit empty fields if they would otherwise occur *many* times in the output, and it's somewhat clear from the file structure that these fields basically exist. `errors` in the tree are a good example: That field occurs many times despite being empty in most cases, and it's somewhat clear from the `has_errors` field that *if* there were errors, they would be listed. For `dependencies` the rationale would be that there almost certainly are some dependencies, so it's clear the field exists, but for the many leaves in the tree it's ok to omit the empty field.

Does that rule make sense? If so, I'd add back serializing of empty `dependencies` for scope, because there usually are not many scopes, and maybe also add back empty `errors` in scan summaries.

File: model/src/main/kotlin/ProjectAnalyzerResult.kt (hunk 2 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/ProjectAnalyzerResult.kt]: It's deserialized in some tests, so I liked to have it here for consistency.

**Comment by mnonnenmacher** [comment]
I would modify that rule to be more of a case-to-case decision: "Does the empty field carry valuable information?" With this applied it would be:

* `errors`: Not serialized, because `has_errors` already carries that information. Also for the `errors` map in `AnalyzerResult`.
* `curations`: Not serialized, because the concept of curations is only important, if they are applied.
* `license_findings`: Serialized, because it is an integral part of the scan result, and it is important to clearly state that no licenses could be found.
* `dependencies` in `Scope`: Serialized, because containing dependencies is the major reason for a scope to exist.
* `data`: Not serialized, because it is not used by ORT itself, but only potentially by other tools.

File: analyzer/src/funTest/assets/projects/synthetic/gradle-all-dependencies-expected-result-with-curations.yml (hunk 1 above)
Comment (sschuberth) [reply analyzer/src/funTest/assets/projects/synthetic/gradle-all-dependencies-expected-result-with-curations.yml]: Ok, I just tried to make that more clear in https://github.com/heremaps/oss-review-toolkit/pull/946.

**Comment by sschuberth** [comment]
Sounds good to me, please make the necessary changes. I'd also actually like to see the rationales for not serializing as code comments whenever we have a `@JsonInclude(JsonInclude.Include.NON_EMPTY)` statement, what do you think?

**Comment by mnonnenmacher** [comment]
@sschuberth Done, will add explanations to the already existing `@JsonInclude`s in a separate PR to not delay this one.

File: model/src/main/kotlin/ProjectAnalyzerResult.kt (hunk 2 above)
Comment (sschuberth) [reply model/src/main/kotlin/ProjectAnalyzerResult.kt]: Maybe more concise: "Do not serialize if empty for consistency with the error properties in other classes, even if this class is not serialized as part of an [OrtResult]."

File: model/src/main/kotlin/PackageReference.kt [hunk 3]
```diff
@@ -37,11 +37,16 @@ data class PackageReference(
          * The list of references to packages this package depends on. Note that this list depends on the scope in
          * which this package reference is used.
          */
-        val dependencies: SortedSet<PackageReference>,
+        // Do not serialize if empty to reduce the size of the result file.
```
Comment (sschuberth) [inline model/src/main/kotlin/PackageReference.kt]: Maybe add: "There likely still are many cases where this is not empty, so it is obvious that dependencies are properly recognized." (Or something like that.)

Comment (mnonnenmacher) [reply model/src/main/kotlin/PackageReference.kt]: Both done.

**Comment by sschuberth** [comment]
One last thing, could we but your great write-up in https://github.com/heremaps/oss-review-toolkit/pull/939#issuecomment-428473277 as the commit message?

**Comment by mnonnenmacher** [comment]
@sschuberth Done.

**Pull request merged**